Plugins/ExaTrkX/include/Acts/Plugins/ExaTrkX/detail/ConnectedComponents.cuh

```diff
@@ -0,0 +1,173 @@
+// This file is part of the ACTS project.
+//
+// Copyright (C) 2016 CERN for the benefit of the ACTS project
+//
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+#pragma once
+
+#include "Acts/Plugins/ExaTrkX/detail/CudaUtils.cuh"
+
+#include <thrust/execution_policy.h>
+#include <thrust/scan.h>
+
+namespace Acts::detail {
+
+template <typename T>
+__device__ void swap(T &a, T &b) {
+  T tmp = a;
+  a = b;
+  b = tmp;
+}
+
+/// Implementation of the FastSV algorithm as shown in
+/// https://arxiv.org/abs/1910.05971
+template <typename TEdge, typename TLabel>
+__global__ void labelConnectedComponents(std::size_t numEdges,
+                                         const TEdge *sourceEdges,
+                                         const TEdge *targetEdges,
+                                         std::size_t numNodes, TLabel *labels,
+                                         TLabel *labelsNext) {
+  // Currently this kernel works only with 1 block
+  assert(gridDim.x == 1);
+
+  for (std::size_t i = threadIdx.x; i < numNodes; i += blockDim.x) {
+    labels[i] = i;
+    labelsNext[i] = i;
+  }
+
+  __shared__ bool changed;
+
+  do {
+    changed = false;
+    __syncthreads();
```

Comment:
Between loop iterations you now have two synchronization points back-to-back.

If I may suggest a cleaner way of doing this, I'd make `changed` a thread-local variable, remove this synchronization point, and then turn the while condition into:

```cpp
} while (__syncthreads_or(changed));
```

Reply:
Pretty good, I didn't new of `__syncthreads_or`!